### PR TITLE
Allow to configure Terraform State

### PR DESCRIPTION
**Pull request by peter-jozsa**
The goal of this PR is to introduce new configuration variables for Terraform State management feature of GitLab. The new params let users:
- enable/disable Terraform State management
- override the default local storage path
- enable/disable Object Store for terraform states
- setup an `AWS` bucket for Object Store
- setup a `GCS` bucket for Object Store

### CODE REVIEW

**Comment by peter-jozsa** [comment]
@sachilles @sameersbn is there any chance this PR will be accepted in the near future?

**Comment by sachilles** [comment]
@peter-jozsa I think so. But I'll need some time to understand the details in your PR (and get warm with Terraform).

**Comment by sachilles** [comment]
@peter-jozsa Sorry for my late reply. I will include your PR before the #2465 . However, I will not be able to check it until the weekend or next week.

**Comment by peter-jozsa** [comment]
alright, thank you! anyway in the end we ended up not using terraform state management but we are planning to use it in the near future

**Comment by sachilles** [comment]
Hi, I think that for now I will rather update GitLab to stay up to date with the upstream.

**Comment by ymazzer** [comment]
Hi, this feature would be wonderful for people using terraform.

As of now, the state is stored using the tmp local folder which does not seem very safe.

Any change to merge it soon? Or at least to change default terraform state storage location to something like `/home/git/data/terraform_state`?

**Comment by peter-jozsa** [comment]
hi 👋 
I rebased my modifications on current `master`. Hopefully we can speed things up here.

**Comment by sachilles** [comment]
Hi,

I'll do my very best to understand the terraform configuration.

**Comment by Syphon83** [comment]
> Hi,
> 
> I'll do my very best to understand the terraform configuration.

This feature would be great! 
Terraform files should follow the same flow for the objects storage as uploads or artifacts files. 

Hope to see this merged soon! Thank you for your work!

**Comment by paolomainardi** [comment]
@sachilles sorry for chasing, are you planning to merge this one for the next release ?

**Comment by sachilles** [comment]
Hi, I'm planning to merge this PR. 

Please be patient. I think I will get to the final review tonight.

File: README.md [hunk 1]
```diff
@@ -1218,6 +1218,62 @@ Google service account. Defaults to `$GITLAB_OBJECT_STORE_CONNECTION_GOOGLE_CLIE
...
+
+Configure this for an compatible AWS host like minio. Defaults to `$AWS_HOST`
+
+##### `GITLAB_TERRAFORM_STATE_OBJECT_STORE_CONNECTION_AWS_ENDPOINT`
+
+AWS Endpoint like `http://127.0.0.1:9000`. Defaults to `$AWS_ENDPOINT`
+
+##### `GITLAB_TERRAFORM_STATE_OBJECT_STORE_CONNECTION_AWS_PATH_STYLE`
+
+Changes AWS Path Style to 'host/bucket_name/object' instead of 'bucket_name.host/object'. Defaults to `AWS_PATH_STYLE`
```
Comment (sachilles) [inline README.md:1263]: Please change `AWS_PATH_STYLE` to `$AWS_PATH_STYLE`

File: README.md [hunk 2]
```diff
@@ -1218,6 +1218,62 @@ Google service account. Defaults to `$GITLAB_OBJECT_STORE_CONNECTION_GOOGLE_CLIE
...
+
+Enables Object Store for Terraform state that will be remote stored. Defaults to `false`
+
+##### `GITLAB_TERRAFORM_STATE_OBJECT_STORE_REMOTE_DIRECTORY`
+
+Bucket name to store the Terraform state. Defaults to `terraform_state`
+
+##### `GITLAB_TERRAFORM_STATE_OBJECT_STORE_CONNECTION_PROVIDER`
+
+Connection Provider for the Object Store. (`AWS` or `Google`) Defaults to `$GITLAB_OBJECT_STORE_CONNECTION_PROVIDER` (`AWS`)
```
Comment (sachilles) [inline README.md:1239]: Please change this sentence to: 

Connection Provider for the Object Store (`AWS` or `Google`).  Defaults to `$GITLAB_OBJECT_STORE_CONNECTION_PROVIDER` (i.e. `AWS`).

**Review by sachilles — APPROVED** [review]
Thanks for this very valuable post and sorry for the long wait (and subsequent rework). Nevertheless, could you include the small changes in the documentation?

**Pull request merged**